src/java/org/jivesoftware/openfire/roster/Roster.java

```diff
@@ -1153,4 +1162,21 @@ public void readExternal(ObjectInput in) throws IOException, ClassNotFoundExcept
         ExternalizableUtil.getInstance().readExternalizableMap(in, rosterItems, getClass().getClassLoader());
         ExternalizableUtil.getInstance().readStringsMap(in, implicitFrom);
     }
+

```

Comment:
I'm wondering if using a hash of the roster would be a suitable, less complex, less error prone alternative to maintaining manual bookkeeping of roster versions. By re-calculating a hash (implementing hashCode and equals for the Roster class, perhaps), the 'version' of the roster becomes a natural property of the roster. The chances of hash collisions are surely negligible, while the overhead of (re)calculation can probably be solved by a lazy-loaded field.

Reply:
As Dave (an author of the original specification) mentions: "XEP-0237 details exactly that as a sane implementation strategy: https://xmpp.org/extensions/xep-0237.html#impl-exactmatch "